rfcs/20200826-keras-nlp-scoping-design.md

```diff
@@ -0,0 +1,108 @@
+In recent years, Transformer-based models have become the foundation of many NLP workflows.
+These workflows tend to reuse similar components, for which in some cases third-party packages
+have been developed by the open-source community.
+
+These third-party solutions are not always kept up to date or up to the same quality standards as core Keras.
+They also raise the issue of API standardization.
+
+To fix this, we want machine learning engineers to have access to a standard Keras-native,
+optimized, and well-tested set of components to build their Transformer-based (and beyond) NLP workflows.
+
+This provides key user benefits:
+
+- The package would be first-party and thus always up to date with modern best practices.
+- High code quality and testing standards and strict quality control: same level of trust as core Keras
+- A shared API standard across the community
+- Ability for the open-source community to build more advanced solutions *on top* of this package instead of reinventing it
+- Ability for research scientists to benefit from subclassing and customizing base components to quickly test new research ideas
+
+## Design Proposal
+
+`keras-nlp` will include most standard Transformer-based modules, specifically:
+
+- Keras layer components such as Transformer encoder and decoder blocks.
+- Keras task components such as masked language, span labeler and named entity recognition.
+- Tensorflow operations such as beam search.
+- Keras optimizer utilities such as learning rate schedules widely used.
+- Data loader and preprocessing for different dataset, such as SQUAD, GLUE.
+
+## Success criteria for keras-nlp
+
+- Reusable and standardized components that cover the above
+- Easy-to-use API
+- Models run on CPU/GPU/TPU seamlessly
+- State of the art performance
+- Models can be readily deployed to production
+
+## Boundaries between keras-nlp and tf.text
+
+- `tf.text` will contain all pre-processing operations, such as WordPiece Tokenizer, n-grams, that handles strings.
+- `keras-nlp` will contain modeling components that cover workflows past the tokenization stage.
+
+## Boundaries between `keras-nlp` and TensorFlow Addons:
+
+- Highly experimental modeling, layers, losses, etc, live in Addons (e.g. newly published research code).
+- Components from Addons will graduate to Model Garden, given they get sufficient usage,
+and given that they work on CPU/GPU/TPU. The API interface will remain experimental for a short time after graduation,
+so as to leave us the option to make changes based on user feedback.
+
+## Boundaries between keras-nlp and Model Garden
+
+- End to end modeling workflow and model specific details live in Model Garden
+- Model garden will re-use most of the building blocks from keras-nlp
+- Components from Model Garden can graduate to keras-nlp, given they get sufficient usage,
+and given that they work on CPU/GPU/TPU. The API interface should remain stable after graduation.
```

Comment:
@tanzhenyu per discussion at the SIG Addons meeting could we update this to say modular components should not be housed in model garden, and that with sufficient usage/reliability Addons components will be graduated to keras/tf-core as needed/